README.rst
kotti_backend
*************

This plugin turns Kotti CMS into a private content administration area.
Basically ``kotti_backend`` let you achieve the same goals described here (`Kotti CMS intranet`_).

What ``kotti_backend`` does:

1. define and load a new ``private_workflow``. If you are installing kotti_backend on
   an existing Kotti installation you'll need to call ``kotti-reset-workflow yourinifile.ini --purge-existing``
   (be extremely **careful** because all your published contents will be converted to the
   private state)

2. apply the above workflow to ``File`` and ``Image`` (they are workflowless by default)

3. set a default fallback permission to all views

4. override the default populator with a custom ACL (no allow view for everyone)

|build status|_
|code coverage|_

`Find out more about Kotti`_

Development happens at https://github.com/Kotti/kotti_backend

.. |build status| image:: https://secure.travis-ci.org/Kotti/kotti_backend.png?branch=master
.. _build status: http://travis-ci.org/Kotti/kotti_backend
.. _Find out more about Kotti: http://pypi.python.org/pypi/Kotti
.. _Kotti CMS intranet: http://davidemoro.blogspot.it/2015/02/kotti-cms-intranet.html
.. |code coverage| image:: http://codecov.io/github/Kotti/kotti_backend/coverage.svg?branch=master
.. _code coverage: http://codecov.io/github/Kotti/kotti_backend?branch=master


Setup
=====

To enable the extension in your Kotti site, activate the configurator::

    kotti.configurators =
        kotti_backend.kotti_configure

    kotti.use_workflow = kotti_backend:workflows/private_workflow.zcml

If you are going to install ``kotti_backend`` on an already existing site you'll need to reset the
workflow: read the docs and backup first. See `Kotti security`_

If you are going to use ``kotti_backend`` with `kotti_frontend`_ (public website decoupled from the private
content administration area), you can add the following additional options in order to see a "Goto frontend"
action in your edit links::

    kotti_backend.goto_frontend = 1                                 # the script_name (by default /cms) will be wiped out
    kotti_backend.not_publishable_types = TYPENAME

or if you want have a custom url for your frontend you can specify a custom frontend url::

    kotti_backend.frontend_url = http://cms.yourproject.com         # or if you have a custom frontend url you can specify it

where:

* ``kotti_backend.goto_frontend`` adds a "Goto frontend" action in your edit links pointing to the public website

* ``kotti_backend.not_publishable_types`` alters the default "goto frontend" link pointing to the first
  publishable parent. This is useful if you are using not directly publishable contents like portles or
  action links you don't want expose on the frontend. See `kotti_boxes`_ or  `kotti_actions`_.

* ``kotti_backend.frontend_url`` let you use a custom frontend url

.. _Kotti security: http://kotti.readthedocs.org/en/latest/developing/basic/security.html
.. _kotti_frontend: https://github.com/Kotti/kotti_frontend
.. _kotti_boxes: https://github.com/Kotti/kotti_boxes
.. _kotti_actions: https://github.com/Kotti/kotti_actions

Development
===========

Contributions to kotti_backend are highly welcome.
Just clone its `Github repository`_ and submit your contributions as pull requests.

.. _tracker: https://github.com/Kotti/kotti_backend/issues
.. _Github repository: https://github.com/Kotti/kotti_backend


CHANGES.rst
History
=======

0.3.0 (unreleased)
------------------

- Fix problem with workflows (changed zcml format)

- Fix Python 3 compatibility and drop Python 2.7 compatibility

- Fix compatibility with new kotti versions (kotti_image)

- Add tox for test setup

- Image support is optional (you can use kotti_backend with
  no kotti_image installed)


0.2.0 (2015-07-08)
------------------

- Add optional "Goto frontend" edit links on backend


0.1.4 (2015-07-02)
------------------

- Add more tests

- Add optional pyramid.includes setting ``kotti_backend.views.override_root_view```.
  Include it if you want to override the root with the ``@@contents`` view


0.1.3 (2015-06-25)
------------------

- Add important note on README

0.1.2 (2015-06-25)
------------------

- Add support for third party plugins with decoupled frontend (public view permission ``pview``)


0.1.1 (2015-06-25)
------------------

- No more automatic override for ``kotti.use_workflow`` ini setting.
  Now you have to add ``kotti.use_workflow = kotti_backend:workflows/private_workflow.zcml``


0.1.0 (2015-06-22)
------------------

- Create package with ``pcreate -s kotti kotti_backend``.


kotti_backend/__init__.py
# -*- coding: utf-8 -*-

"""
Created on 2015-06-22
:author: Davide Moro (davide.moro@gmail.com)
"""
from zope.interface import implementer
from pyramid.settings import asbool
from pyramid.i18n import TranslationStringFactory
from kotti.interfaces import IDefaultWorkflow
from kotti.resources import (
    File,
    Content,
    )
from kotti.util import Link
try:
    from kotti_image.resources import Image
except ImportError:
    HAS_IMAGE = False
else:
    HAS_IMAGE = True


_ = TranslationStringFactory('kotti_backend')


def kotti_configure(settings):
    """ Add a line like this to you .ini file::

            kotti.configurators =
                kotti_backend.kotti_configure

        to enable the ``kotti_backend`` add-on.

    :param settings: Kotti configuration dictionary.
    :type settings: dict
    """

    settings['pyramid.includes'] += ' kotti_backend'
    settings['kotti.configurators'] += ' kotti_image.kotti_configure'
    settings['kotti.populators'] = 'kotti_backend.populate.populate'

    if asbool(settings.get('kotti_backend.goto_frontend')):
        frontend_link = Link(
            name='goto_frontend',
            title=_('Goto frontend'),
            target='_blank',
            )
        Content.type_info.edit_links.append(frontend_link)


def includeme(config):
    """ Don't add this to your ``pyramid_includes``, but add the
    ``kotti_configure`` above to your ``kotti.configurators`` instead.

    :param config: Pyramid configurator object.
    :type config: :class:`pyramid.config.Configurator`
    """

    # Set a default permission.
    # If you want to bypass the default permission for certain views,
    # you can decorate them with a special permission
    # (``NO_PERMISSION_REQUIRED`` from ``pyramid.security`` which indicates
    # that the view should always be executable by entirely anonymous users,
    # regardless of the default permission.
    config.set_default_permission('view')

    # Assign the default workflow for files and images
    if HAS_IMAGE:
        implementer(IDefaultWorkflow)(Image)
    implementer(IDefaultWorkflow)(File)

    base_includes = ('kotti_backend.views.goto_frontend',)
    for include in base_includes:
        config.include(include)

    # translations
    config.add_translation_dirs('kotti_backend:locale')
